### PR TITLE
Add extra version of py-ipywidgets

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 1]
```diff
@@ -12,6 +12,7 @@ class PyIpywidgets(PythonPackage):
     homepage = "https://github.com/ipython/ipywidgets"
     url      = "https://github.com/ipython/ipywidgets/archive/5.2.2.tar.gz"
 
+    version('7.4.2', sha256='f156165e8a855ed862fdf48e72700bdcd6956d089a2018c5b36d358255d45b2b')
     version('5.2.2', sha256='d61ab8bb12b90981a3a6010429816d70eaa041e622043207bcb74239b664d4f3')
 
     depends_on('python@2.7:2.8,3.3:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('python@2.7:2.8,3.3:', type=('build', 'run'))
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: There's also some changes to the dependencies. You'll have to see which version added these:
```python
install_requires = setuptools_args['install_requires'] = [                      
    'ipykernel>=4.5.1',                                                         
    'traitlets>=4.3.1',                                                         
    # Requiring nbformat to specify bugfix version which is not required by     
    # notebook.                                                                 
    'nbformat>=4.2.0',                                                          
    # TODO: Dynamically add this dependency                                     
    # only if notebook 4.x is installed in this                                 
    # interpreter, to allow ipywidgets to be                                    
    # installed on bare kernels.                                                
    'widgetsnbextension~=3.4.0',                                                
]                                                                               
                                                                                
extras_require = setuptools_args['extras_require'] = {                          
    ':python_version<"3.3"' : ['ipython>=4.0.0,<6.0.0'],                        
    ':python_version>="3.3"': ['ipython>=4.0.0'],                               
    'test:python_version=="2.7"': ['mock'],                                     
    'test': ['nose'],                                                           
}                     
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: Don't forget to add these dependencies.

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 2]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
     homepage = "https://github.com/ipython/ipywidgets"
     url      = "https://github.com/ipython/ipywidgets/archive/5.2.2.tar.gz"
 
+    version('7.4.2', sha256='f156165e8a855ed862fdf48e72700bdcd6956d089a2018c5b36d358255d45b2b')
     version('5.2.2', sha256='d61ab8bb12b90981a3a6010429816d70eaa041e622043207bcb74239b664d4f3')
 
-    depends_on('python@2.7:2.8,3.3:')
-    depends_on('py-ipython@4.0.0:', type=('build', 'run'))
+    depends_on('python@2.7:2.8,3.3:', type=('build', 'run'))
+    depends_on('py-ipython@4.0.0:', type=('build', 'run'), when='^python@3.3:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-ipython@4:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 3]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
+    depends_on('py-ipython@4.0.0:5.999.999', type=('build', 'run'), when='^python@:3.2.999')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-ipython@4:5', type=('build', 'run'), when='^python@:3.2')
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 4]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
     depends_on('py-ipykernel@4.2.2:', type=('build', 'run'))
+    depends_on('py-ipykernel@4.5.1:', type=('build', 'run'), when='@6.0.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-ipykernel@4.5.1:', type=('build', 'run'), when='@6:')
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 5]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
     depends_on('py-traitlets@4.2.1:', type=('build', 'run'))
+    depends_on('py-traitlets@4.3.1:', type=('build', 'run'), when='@6.0.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-traitlets@4.3.1:', type=('build', 'run'), when='@6:')
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 6]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
+    depends_on('py-nbformat@4.2.0:', type=('build', 'run'), when='@6.0.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-nbformat@4.2.0:', type=('build', 'run'), when='@6:')
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 7]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
+    depends_on('py-widgetsnbextension@3.4.0:3.4.999', type=('build', 'run'), when='@6.0.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]:
```suggestion
    depends_on('py-widgetsnbextension@3.4.0:3.4.999', type=('build', 'run'), when='@6:')
```

File: var/spack/repos/builtin/packages/py-ipywidgets/package.py [hunk 8]
```diff
@@ -12,9 +12,17 @@ class PyIpywidgets(PythonPackage):
...
-    depends_on('py-ipython@4.0.0:', type=('build', 'run'))
+    depends_on('python@2.7:2.8,3.3:', type=('build', 'run'))
+    depends_on('py-ipython@4:', type=('build', 'run'))
+    depends_on('py-ipython@4:5', type=('build', 'run'), when='^python@:3.2')
     depends_on('py-ipykernel@4.2.2:', type=('build', 'run'))
+    depends_on('py-ipykernel@4.5.1:', type=('build', 'run'), when='@6:')
     depends_on('py-traitlets@4.2.1:', type=('build', 'run'))
+    depends_on('py-traitlets@4.3.1:', type=('build', 'run'), when='@6:')
+    depends_on('py-nbformat@4.2.0:', type=('build', 'run'), when='@6:')
+    depends_on('py-widgetsnbextension@3.4.0:3.4.999', type=('build', 'run'), when='@6:')
```
Comment (glennpj) [inline var/spack/repos/builtin/packages/py-ipywidgets/package.py]: Is there a PR to update py-widgetsnbextension?

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: Now there is: #15105

Comment (glennpj) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: #15105 is for version 3.3.0 but the dependency above is for 3.4.

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: Added 3.5.1

Comment (glennpj) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: You will need 3.4.x to match the dependency specified for ipywidgets.
```
depends_on('py-widgetsnbextension@3.4.0:3.4.999', type=('build', 'run'), when='@6:')
```

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-ipywidgets/package.py]: And 3.4.2

**Pull request merged**